feat(hmac-auth): Add validate request body for hmac auth plugin (#5038)
apache/apisix: docs/en/latest/plugins/hmac-auth.md

| keep_headers      | boolean       | optional    | false         | [ true, false ]                             | Whether it is necessary to keep the request headers of `X-HMAC-SIGNATURE`, `X-HMAC-ALGORITHM` and `X-HMAC-SIGNED-HEADERS` in the http request after successful authentication. true: means to keep the http request header, false: means to remove the http request header.   |
| encode_uri_params | boolean       | optional    | true          | [ true, false ]                             | Whether to encode the uri parameter in the signature, for example: `params1=hello%2Cworld` is encoded, `params2=hello,world` is not encoded. true: means to encode the uri parameter in the signature, false: not to encode the uri parameter in the signature.               |
| validate_request_body | boolean  | optional   | false         | [ true, false ]                             | Whether to check request body. |

> | max_req_body     | number        | optional   | 512KB         |                                             | Max allowed body size. |

Reviewer: Better to use a number instead of string in the doc's default value cell

| --------- | -------------------------------------------- |
| SIGNATURE | 8XV1GB7Tq23OJcoz6wjqTs4ZLxr9DiLoY4PxzScWGYg= |

### Request body checking

> When `validate_request_body` is assigned to `true`, the plugin will check the request body. The plugin will calculate the hmac-sha value of the request body，and check against the `X-HMAC-DIGEST` header.

Reviewer: Better to mention the max_req_body limitation

Author: I add a **note** to that

```
X-HMAC-DIGEST: base64(hmac-sha(<body>))
```

> When there is no request body, the `X-HMAC-DIGEST` header can be omitted. If you want to send request with this header whether the body is empty or not, you can set `X-HMAC-DIGEST` value to the hmac-sha of empty string.

Reviewer: ```suggestion
When there is no request body, the `X-HMAC-DIGEST` header can be omitted. If you want to send a request with this header when the body is missing, you can set `X-HMAC-DIGEST` value to the hmac-sha of empty string.
```

Author: done

| --------- | -------------------------------------------- |
| SIGNATURE | 8XV1GB7Tq23OJcoz6wjqTs4ZLxr9DiLoY4PxzScWGYg= |

### Request body checking

When `validate_request_body` is assigned to `true`, the plugin will check the request body. The plugin will calculate the hmac-sha value of the request body，and check against the `X-HMAC-DIGEST` header.

```
X-HMAC-DIGEST: base64(hmac-sha(<body>))
```

When there is no request body, the `X-HMAC-DIGEST` header can be omitted. If you want to send a request with this header when the body is missing, you can set `X-HMAC-DIGEST` value to the hmac-sha of empty string.

> **Note:** The plugin will load the request body to memory to calculate the digest of the request body, which wight cause high memory consumption with large bodies. You can limit the max allowed body size by the configuration of `max_req_body`(default=512KB), request body larger than that will be rejected.

Reviewer: ```suggestion
**Note:** The plugin will load the request body to memory to calculate the digest of the request body, which might cause high memory consumption with large bodies. You can limit the max allowed body size by the configuration of `max_req_body`(default=512KB), request body larger than that will be rejected.
```

<html lang="cn">
```

### Enable request body checking

```shell
$ curl -X "POST" "http://localhost:9080/index.html?age=36&name=james" \
     -H 'X-HMAC-ACCESS-KEY: zyedu-hmac-01' \
     -H 'X-HMAC-SIGNATURE: ivlwjZPoVdSVvdSSM4drEFk9q9HS2jeJ5cAN9JffmdA=' \
     -H 'X-HMAC-ALGORITHM: hmac-sha256' \
     -H 'Date: Tue, 24 Aug 2021 03:19:21 GMT' \
     -H 'X-HMAC-SIGNED-HEADERS: User-Agent;Digest' \
     -H 'User-Agent: curl/7.29.0' \
     -H 'Digest: L9b/+QMvhvnoUlSw5vq+kHPqnZiHGl61T8oavMVTaC4=' \
     -H 'Content-Type: text/plain; charset=utf-8' \
     -d "{\"hello\":\"world\"}"

> HTTP/1.1 200 OK

Reviewer: Better to only show the first line? APISIX 2.2 doesn't support this feature. We can avoid confusing users by only show the first line.

Author: Got it
